pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -221,17 +221,33 @@ func podToEndpointAddressForService(svc *v1.Service, pod *v1.Pod) (*v1.EndpointA
 		// In a legacy cluster, the pod IP is guaranteed to be usable
 		endpointIP = pod.Status.PodIP
 	} else {
-		ipv6Service := endpointutil.IsIPv6Service(svc)
-		for _, podIP := range pod.Status.PodIPs {
-			ipv6PodIP := utilnet.IsIPv6String(podIP.IP)
-			if ipv6Service == ipv6PodIP {
-				endpointIP = podIP.IP
-				break
+		//feature flag enabled and pods may have multiple IPs
+		ipv6Service := false
+
+		if len(svc.Spec.ClusterIP) > 0 && svc.Spec.ClusterIP != v1.ClusterIPNone {
+			// headful service. detect via service clusterIP
+			ipv6Service = utilnet.IsIPv6String(svc.Spec.ClusterIP)
+			for _, podIP := range pod.Status.PodIPs {
+				ipv6PodIP := utilnet.IsIPv6String(podIP.IP)
+				if ipv6Service == ipv6PodIP {
+					endpointIP = podIP.IP
+					break
+				}
 			}
+		} else {
+			// THE BELOW IS VALID ONLY FOR CLASSIC ENDPOINT CONTROLLER
+			// Since this is a headless service we use podIP as-is.
+			// The old singular status field always holds IP of the primary family of the cluster.
```

Comment:
`pod.Status.PodIP` is not guaranteed to be of the same family as the primary service IP. The KEP seems to still claim that it is (https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md#default-pod-ip-selection) but we had agreed in the last go-round that that's incorrect, because there's nothing that forces the network plugin to use the same family ordering for pod IPs as the apiserver uses for service IPs, so it's possible that services have IPv6,IPv4 but pods have IPv4,IPv6.

At any rate, even if it _was_ guaranteed to be the primary service IP, it still doesn't make sense to ignore `svc.Spec.IPFamilies[0]` here. If the service says that it's supposed to be single-stack IPv6, then we should use the IPv6 Pod IP.

Reply:
Good point. So I modified our approach. If there are IPFamilies we use them, if not (old api-server) we fall back to the above solution. further closing the gap.